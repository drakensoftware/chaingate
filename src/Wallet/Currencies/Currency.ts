import Decimal from 'decimal.js'
import {Api} from 'chaingate-client'
import {Phrase, PrivateKey, PrivateKeySource, PublicKey, Seed} from '../Keys'

export class NotEnoughFundsError extends Error {
    constructor(currency: Currency, missingFunds: CurrencyAmount) {
        super(`Wallet does not have enough funds. ${missingFunds.baseAmount} ${missingFunds.baseSymbol} needed.`)
        if (Error.captureStackTrace) Error.captureStackTrace(this, NotEnoughFundsError)
        this.name = this.constructor.name
    }
}

export class CannotDerive extends Error {
    constructor() {
        super('Derivation path are not supported (Are you using an imported private key?)')
        if (Error.captureStackTrace) Error.captureStackTrace(this, CannotDerive)
        this.name = this.constructor.name
    }
}

export class CannotParseAmount extends Error {
    constructor(amountStr: string) {
        super(`Invalid amount: ${amountStr}`)
        if (Error.captureStackTrace) Error.captureStackTrace(this, CannotDerive)
        this.name = this.constructor.name
    }
}

export type FeeGrade = 'low' | 'normal' | 'high' | 'maximum'

export abstract class ConfirmedTransaction {
    abstract readonly txId: string

    abstract isConfirmed(): Promise<boolean>
    waitToBeConfirmed(): Promise<void> {
        // eslint-disable-next-line no-async-promise-executor
        return new Promise(async (resolve) => {
            // First check before starting the interval
            if (await this.isConfirmed()) {
                resolve()
            } else {
                const intervalId = setInterval(async () => {
                    if (await this.isConfirmed()) {
                        clearInterval(intervalId)
                        resolve()
                    }
                }, 10000)
            }
        })
    }
}

export abstract class PreparedTransaction {
    abstract readonly possibleFees: { [key in FeeGrade]: Fee }
    abstract confirm(fee: string | Fee): Promise<ConfirmedTransaction>
}

export class CurrencyAmount {
    readonly baseAmount: Decimal
    readonly baseSymbol: string

    readonly minimalUnitAmount: Decimal
    readonly minimalUnitSymbol: string

    constructor(currency: Currency, baseAmount: Decimal){
        this.baseAmount = baseAmount
        this.minimalUnitAmount = baseAmount.mul( (new Decimal(10)).pow(currency.currencyInfo.decimals))

        this.baseSymbol = currency.currencyInfo.symbol
        this.minimalUnitSymbol = currency.currencyInfo.minimalUnitSymbol
    }
}

export abstract class Fee {}
export abstract class FeeInitializer {}
export type PossibleFees<T> = {[key in FeeGrade]: T}

export type CurrencyInfo = {
    id: string
    name: string
    defaultDerivationPath: string
    svgLogoUrl: string,
    symbol: string,
    minimalUnitSymbol: string,
    decimals: number
}

export abstract class Currency{
    protected readonly api: Api
    public readonly currencyInfo: CurrencyInfo

    protected publicKey: PublicKey | null
    protected privateKeySource: PrivateKeySource

    protected constructor(currencyInfo: CurrencyInfo, api: Api, privateKeySource: PrivateKeySource) {
        this.currencyInfo = currencyInfo
        this.api = api
        this.privateKeySource = privateKeySource
        this._derivationPath = currencyInfo.defaultDerivationPath
    }

    async getPrivateKey() {
        if (this.privateKeySource instanceof PrivateKey) return this.privateKeySource
        else if (this.privateKeySource instanceof Seed || this.privateKeySource instanceof Phrase) {
            return this.privateKeySource.derive(this.getDerivationPath())
        }
    }

    async initializePublicKey(){
        this.publicKey = await (await this.getPrivateKey()).getPublicKey()
    }

    amount(amountStr: string) : CurrencyAmount{
        const regex = /^(\d+(\.\d+)?) ?([A-Za-z]+)?$/
        const match = amountStr.match(regex)
        if (!match) throw new CannotParseAmount(amountStr)

        const amount = new Decimal(match[1])

        if (match[3]) {
            let currencySymbol = match[3].toLowerCase()
            if(currencySymbol.endsWith('s')) currencySymbol = currencySymbol.substring(currencySymbol.length - 1) //Remove plural -s

            if (currencySymbol === this.currencyInfo.symbol.toLowerCase() ||
                currencySymbol === this.currencyInfo.name.toLowerCase()) {
                return new CurrencyAmount(this, amount)
            } else if (currencySymbol === this.currencyInfo.minimalUnitSymbol.toLowerCase()) {
                return new CurrencyAmount(this, amount.mul((new Decimal(10)).pow(this.currencyInfo.decimals)))
            } else {
                throw new CannotParseAmount(amountStr)
            }
        }

        return new CurrencyAmount(this, amount)
    }


    abstract fee(fee: unknown): Fee

    abstract getAddress(): Promise<string>

    abstract getBalance(address?: string): Promise<{confirmed: CurrencyAmount, unconfirmed: CurrencyAmount}>

    abstract prepareTransfer(toAddress: string, amount: CurrencyAmount): Promise<PreparedTransaction>

    // #region Derivation Paths
    private _derivationPath: string | null

    getDerivationPath(){
        return this._derivationPath
    }

    async setDerivationPath(newDerivationPath: string){
        this._derivationPath = newDerivationPath
        await this.initializePublicKey()
    }
    // #endregion
}
