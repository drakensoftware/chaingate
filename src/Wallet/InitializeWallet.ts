import {PhaseNumOfWords, PhraseGenerator, PhraseLanguage} from '../Utils/PhraseGenerator'
import {PhraseWallet} from './PhraseWallet'
import {ChainGateClient} from 'chaingate-client'
import {Phrase} from './Keys'
import {decryptKeystoreV1, isKeystoreV1} from './Crypto/KeystoreDecoder/KeystoreV1'
import {decryptKeystore} from './Crypto/KeystoreDecoder/OtherKeystores'
import {ImportedPrivateKey} from './ImportedPrivateKey'
import {PrivateKey} from './Keys'
import IncorrectPassword from './Crypto/KeystoreDecoder/IncorrectPassword'
import {Utils} from '../Utils'
import {HDWallet} from './HDWallet'
import {Seed} from './Keys'
import wif from 'wif'

export class EncodingError extends Error {
    constructor(message: string) {
        super(message)
        if (Error.captureStackTrace) Error.captureStackTrace(this, EncodingError)
        this.name = this.constructor.name
    }
}

export async function create(apiKey: string,
    phraseLanguage: PhraseLanguage = 'english', phraseNumOfWords: PhaseNumOfWords = 12) {
    const phrase = PhraseGenerator.generateNewPhrase(phraseLanguage, phraseNumOfWords)
    return await PhraseWallet.new(new ChainGateClient(apiKey), new Phrase(phrase))
}

export async function fromKeystore(apiKey: string, keystore: string, password: string) {
    //Some keystores store the phrase and others the private key
    try {
        const decodedKeystore = isKeystoreV1(keystore)
            ? await decryptKeystoreV1(keystore, password)
            : await decryptKeystore(keystore, password)

        let phrase
        try {
            phrase = new TextDecoder().decode(decodedKeystore)
        } catch { /* empty */ }

        if (phrase && Phrase.isValidPhrase(phrase)) {
            return PhraseWallet.new(new ChainGateClient(apiKey), new Phrase(phrase))
        } else {
            return ImportedPrivateKey.new(new ChainGateClient(apiKey), new PrivateKey(decodedKeystore))
        }
    } catch (ex) {
        if (ex instanceof IncorrectPassword) throw ex
        throw new EncodingError('Invalid keystore')
    }
}

export async function fromPhrase(apiKey: string, phrase: string) {
    if(!Phrase.isValidPhrase(phrase)) throw new EncodingError('Invalid phrase')
    return PhraseWallet.new(new ChainGateClient(apiKey), new Phrase(phrase))
}

export async function fromSeed(apiKey: string, seed: string | Uint8Array, ){
    if(typeof seed === 'string' && Utils.isHex(seed)) seed = Utils.hexToBytes(seed)
    else if(typeof seed === 'string') throw new EncodingError('Invalid seed')

    return HDWallet.new(new ChainGateClient(apiKey), new Seed(seed))
}

export async function  fromPrivateKey(apiKey: string, privateKey: string | Uint8Array){
    if(typeof privateKey === 'string' && Utils.isHex(privateKey)) privateKey = Utils.hexToBytes(privateKey)
    else if(typeof privateKey === 'string' && Utils.isBase58(privateKey)){
        try{
            privateKey = new Uint8Array(wif.decode(privateKey).privateKey)
        }catch (_ex){
            throw new EncodingError('The string supplied in Wallet Import Format (WIF) is deemed to be invalid')
        }
    }else if(typeof privateKey === 'string') throw new EncodingError('Private key is invalid or the format is unrecognized')

    return ImportedPrivateKey.new(new ChainGateClient(apiKey), new PrivateKey(privateKey))
}
