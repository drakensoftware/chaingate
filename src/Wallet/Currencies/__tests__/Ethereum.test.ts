import {NotEnoughFundsError} from '../Currency'
import * as InitializeWallet from '../../InitializeWallet'

describe('Ethereum', function () {
    it('Generate addresses of different types', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const mnemonic = 'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about'
        const wallet = await InitializeWallet.fromPhrase('abc', mnemonic)
        const ethereum = wallet.currencies.ethereum

        expect(await ethereum.getAddress())
            .toBe('0x9858EfFD232B4033E47d90003D41EC34EcaEda94')
    })

    it('Get address balance', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const mnemonic = 'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about'
        const wallet = await InitializeWallet.fromPhrase('', mnemonic)
        const ethereum = wallet.currencies.ethereum

        expect(await ethereum.getAddress()).toBe('0x9858EfFD232B4033E47d90003D41EC34EcaEda94')

        const balance = await ethereum.getBalance()
        expect(balance.confirmed.baseAmount.toString()).toBe('1.1909e-14')
        expect(balance.unconfirmed.baseAmount.toNumber()).toBe(0)
    })

    it('Transfer some ethereum', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const mnemonic = 'zero sun street chronic online omit buzz critic south sample video drastic base provide cream'
        const wallet = await InitializeWallet.fromPhrase('', mnemonic)
        const ethereum = wallet.currencies.ethereum

        expect(await ethereum.getAddress()).toBe('0xfF9810e98C9c0Fa3DDaB062413464A956aD87b31')

        const balance = await ethereum.getBalance()
        expect(balance.confirmed.baseAmount.toString()).toBe('0.032352785715404185')

        const transaction = await ethereum.prepareTransfer(
            '0x649Ac0b64Fe245783139f44344AD39a49f3B0E32', ethereum.amount('0.01'))
        expect(transaction.possibleFees).toMatchSnapshot()

        await transaction.confirm('normal')
    })

    it('Transfer some ethereum (not enough funds)', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const mnemonic = 'zero sun street chronic online omit buzz critic south sample video drastic base provide cream'
        const wallet = await InitializeWallet.fromPhrase('', mnemonic)
        const ethereum = wallet.currencies.ethereum

        expect(await ethereum.getAddress()).toBe('0xfF9810e98C9c0Fa3DDaB062413464A956aD87b31')

        await expect(async () =>
            await ethereum.prepareTransfer('0x649Ac0b64Fe245783139f44344AD39a49f3B0E32', ethereum.amount('1'))).rejects
            .toThrow(NotEnoughFundsError)
    })

    it('Call smart contract function', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const mnemonic = 'zero sun street chronic online omit buzz critic south sample video drastic base provide cream'
        const wallet = await InitializeWallet.fromPhrase('', mnemonic)
        const ethereum = wallet.currencies.ethereum

        expect(await ethereum.getAddress()).toBe('0xfF9810e98C9c0Fa3DDaB062413464A956aD87b31')

        const result = await ethereum.callSmartContractFunction(
            '0x6b175474e89094c44da98b954eedeac495271d0f',
            '0x70a082310000000000000000000000006E0d01A76C3Cf4288372a29124A26D4353EE51BE'
        )

        expect(result).toMatchSnapshot()
    })

})
