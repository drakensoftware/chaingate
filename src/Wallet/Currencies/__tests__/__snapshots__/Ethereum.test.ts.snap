// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ethereum Call smart contract function 1`] = `"0x0000000000000000000000000000000000000000000000000858898f93629000"`;

exports[`Ethereum Transfer some ethereum 1`] = `
{
  "high": EvmFee {
    "maxFeePerGas": CurrencyAmount {
      "baseAmount": "5e-18",
      "baseSymbol": "ETH",
      "minimalUnitAmount": "5",
      "minimalUnitSymbol": "wei",
    },
    "maxPriorityFeePerGas": CurrencyAmount {
      "baseAmount": "1",
      "baseSymbol": "ETH",
      "minimalUnitAmount": "1000000000000000000",
      "minimalUnitSymbol": "wei",
    },
  },
  "low": EvmFee {
    "maxFeePerGas": CurrencyAmount {
      "baseAmount": "4e-18",
      "baseSymbol": "ETH",
      "minimalUnitAmount": "4",
      "minimalUnitSymbol": "wei",
    },
    "maxPriorityFeePerGas": CurrencyAmount {
      "baseAmount": "0",
      "baseSymbol": "ETH",
      "minimalUnitAmount": "0",
      "minimalUnitSymbol": "wei",
    },
  },
  "maximum": EvmFee {
    "maxFeePerGas": CurrencyAmount {
      "baseAmount": "6e-18",
      "baseSymbol": "ETH",
      "minimalUnitAmount": "6",
      "minimalUnitSymbol": "wei",
    },
    "maxPriorityFeePerGas": CurrencyAmount {
      "baseAmount": "3",
      "baseSymbol": "ETH",
      "minimalUnitAmount": "3000000000000000000",
      "minimalUnitSymbol": "wei",
    },
  },
  "normal": EvmFee {
    "maxFeePerGas": CurrencyAmount {
      "baseAmount": "4e-18",
      "baseSymbol": "ETH",
      "minimalUnitAmount": "4",
      "minimalUnitSymbol": "wei",
    },
    "maxPriorityFeePerGas": CurrencyAmount {
      "baseAmount": "0",
      "baseSymbol": "ETH",
      "minimalUnitAmount": "0",
      "minimalUnitSymbol": "wei",
    },
  },
}
`;
